d2/src/main/java/com/linkedin/d2/balancer/clusterfailout/FailedoutClusterManager.java

```diff
@@ -177,11 +194,51 @@ private void removeClusterWatches(@Nonnull Set<String> clustersToRemove)
     }
     for (String cluster : clustersToRemove)
     {
-      final LoadBalancerStateListenerCallback listener = _clusterListenerCallbacks.remove(cluster);
-      if (listener != null)
+      final PeerWatchState peerWatchState = _peerWatches.remove(cluster);
+      if (peerWatchState == null)
+      {
+        continue;
+      }
+      if (_connectionWarmUpHandler != null)
+      {
+        _log.debug("Cancel pending requests to: {}", cluster);
+        _connectionWarmUpHandler.cancelPendingRequests(cluster);
+      }
+      if (peerWatchState.shouldUnregisterWatches())
```

Comment:
Maybe add a comment here: we only unregister the watches if they do not exist prior to fail out.